plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsLocalIndexExporter.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.exporter;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexResponse;
+import org.opensearch.action.admin.indices.delete.DeleteIndexRequest;
+import org.opensearch.action.bulk.BulkRequest;
+import org.opensearch.action.bulk.BulkResponse;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.action.support.WriteRequest;
+import org.opensearch.action.support.master.AcknowledgedResponse;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Class to export data collected by search query analyzers to a local OpenSearch index
+ * <p>
+ * Internal used within the Query Insight framework
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsLocalIndexExporter<T extends SearchQueryRecord<?>> extends QueryInsightsExporter<T> {
+    private static final Logger log = LogManager.getLogger(QueryInsightsLocalIndexExporter.class);
+    private static final int INDEX_TIMEOUT = 60;
+
+    private final ClusterService clusterService;
+    private final Client client;
+
+    /** The mapping for the local index that holds the data */
+    private final InputStream localIndexMapping;
+
+    public QueryInsightsLocalIndexExporter(
+        ClusterService clusterService,
+        Client client,
+        String localIndexName,
+        InputStream localIndexMapping
+    ) {
```

Comment:
Should the each concrete instance of this class be singleton for latency, cpu or memory types? We are not planning to use multiple local index exporter for same type of information like latency or cpu?

Reply:
One instance of this class would be for one index with the corresponding mapping. For the same type of information, if they are saved to different index based on user's config, different index exporter would be used.